### PR TITLE
Donation refactor 2

### PR DESCRIPTION
**Pull request by asood123**

### CODE REVIEW

File: server/controllers/donations.js [hunk 1]
```diff
@@ -331,9 +332,19 @@ module.exports = (app) => {
           .catch(cb);
       },
 
+      createSubscription: ['getConnectedAccount', (cb) => {
```
Comment (sedubois) [inline server/controllers/donations.js]: would be nice to use promises

File: server/controllers/transactions.js [hunk 2]
```diff
@@ -1,421 +0,0 @@
-/**
```
Comment (sedubois) [inline server/controllers/transactions.js]: great to get rid of this!

File: server/lib/notifications.js [hunk 3]
```diff
@@ -27,7 +27,7 @@ module.exports = (Sequelize, activity) => {
             activity.type
           ],
           GroupId: activity.GroupId,
-          channel: ['gitter', 'slack', 'twitter'],
+          channel: ['gitter', 'slack', 'twitter', 'email'],
```
Comment (sedubois) [inline server/lib/notifications.js]: Why is this added, as nothing is done with it on line 35 of this same file?

File: server/models/Transaction.js [hunk 4]
```diff
@@ -158,6 +159,89 @@ module.exports = function(Sequelize, DataTypes) {
           netAmountInGroupCurrency: this.netAmountInGroupCurrency
         };
       }
+    },
+
+    classMethods: {
+      createFromPayload(payload) {
```
Comment (sedubois) [inline server/models/Transaction.js]: Should use ES6 destructuring syntax `{user, group, transaction, subscription, paymentMethod}` with appropriate default values (if any) instead of the black-box `payload`.

File: server/models/Transaction.js [hunk 5]
```diff
@@ -158,6 +159,89 @@ module.exports = function(Sequelize, DataTypes) {
...
+        }
+        return Transaction.findById(transaction.id, {
+          include: [
+            { model: Sequelize.models.Group },
+            { model: Sequelize.models.User },
+            { model: Sequelize.models.PaymentMethod }
+          ]
+        })
+        // Create activity.
+        .then(transaction => {
```
Comment (sedubois) [inline server/models/Transaction.js]: this obscures the method param which is bad for readability

File: server/models/Transaction.js [hunk 6]
```diff
@@ -158,6 +159,89 @@ module.exports = function(Sequelize, DataTypes) {
...
+                - transaction.paymentProcessorFeeInTxnCurrency)
+              *transaction.txnCurrencyFxRate);
+        } else {
+          // populate netAmountInGroupCurrency for "Add Funds" and Expenses
+          transaction.netAmountInGroupCurrency = transaction.amount*100;
+        }
+        return Transaction.create(transaction);
+      },
+
+      createActivity(transaction) {
```
Comment (sedubois) [inline server/models/Transaction.js]: suggest to destructure param same as method above

File: server/models/Transaction.js [hunk 7]
```diff
@@ -158,6 +159,89 @@ module.exports = function(Sequelize, DataTypes) {
...
+          if (transaction.User) {
+            activityPayload.data.user = transaction.User.info;
+          }
+          if (transaction.PaymentMethod) {
+            activityPayload.data.paymentMethod = transaction.PaymentMethod.info;
+          }
+          return Sequelize.models.Activity.create(activityPayload);
+        })
+        .catch(err => console.error(`Error creating activity of type ${activities.GROUP_TRANSACTION_CREATED} for transaction ID ${transaction.id}`, err))
+        // notify subscribers. TODO: #EmailLibRefactor bring this back once we refactor emailLib.
```
Comment (sedubois) [inline server/models/Transaction.js]: Is it OK to merge this PR without this TODO?

File: server/routes.js [hunk 8]
```diff
@@ -156,7 +155,6 @@ module.exports = (app) => {
   /**
    * Transactions (financial).
    */
-  // TODO remove #postmigration, replaced by GET /groups/:groupid/expenses
   app.get('/groups/:groupid/transactions', aZ.authorizeAccessToGroup({authIfPublic: true}), mw.paginate(), mw.sorting({key: 'createdAt', dir: 'DESC'}), groups.getTransactions); // Get a group's transactions.
 
   // TODO remove #postmigration, replaced by POST /groups/:groupid/expenses
```
Comment (sedubois) [inline server/routes.js]: NB: just realised this might still be used by "add funds"

File: test/groups.routes.test.js [hunk 9]
```diff
@@ -409,26 +408,20 @@ describe('groups.routes.test.js', () => {
     });
 
     beforeEach((done) => {
```
Comment (sedubois) [inline test/groups.routes.test.js]: normally can remove the `done` as well as lines 416-417

File: test/groups.routes.test.js [hunk 10]
```diff
@@ -567,12 +560,14 @@ describe('groups.routes.test.js', () => {
 
       // Create a subscription for PublicGroup.
       beforeEach((done) => {
```
Comment (sedubois) [inline test/groups.routes.test.js]: similar note, I think the `done` can be avoided

File: test/subscriptions.routes.test.js [hunk 11]
```diff
@@ -54,14 +54,16 @@ describe('subscriptions.routes.test.js', () => {
   describe('#getAll', () => {
     // Create transactions for group1.
     beforeEach((done) => {
```
Comment (sedubois) [inline test/subscriptions.routes.test.js]: can normally remove the `done` and lines 65-66?

File: test/subscriptions.routes.test.js [hunk 12]
```diff
@@ -105,16 +107,17 @@ describe('subscriptions.routes.test.js', () => {
     var nocks = {};
 
     beforeEach((done) => {
```
Comment (sedubois) [inline test/subscriptions.routes.test.js]: get rid of `done` and lines 119-120?

**Comment by sedubois** [comment]
OK for me to merge if you're fine with the comments

File: server/controllers/donations.js (hunk 1 above)
Comment (asood123) [inline server/controllers/donations.js]: yes, we'll come back and refactor that after this.

File: server/lib/notifications.js (hunk 3 above)
Comment (asood123) [inline server/lib/notifications.js]: good catch. I wanted to refactor the emailLib and add that through the notification library. I'll do that separately.

File: server/models/Transaction.js (hunk 4 above)
Comment (asood123) [inline server/models/Transaction.js]: done

File: server/models/Transaction.js (hunk 5 above)
Comment (asood123) [inline server/models/Transaction.js]: this one is a little trickier

File: server/models/Transaction.js (hunk 7 above)
Comment (asood123) [inline server/models/Transaction.js]: Let me see if I can do this now that xavier's last change is merged.
